controllers/storageautoscaler/scraper.go

```diff
@@ -0,0 +1,146 @@
+package storageautoscaler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+// StorageAutoscalerScrpaer is the scraper for Prometheus metrics
+type StorageAutoscalerScraper struct {
+	client.Client
+	Log               logr.Logger
+	OperatorNamespace string
+	ScrapeInterval    time.Duration
+	SyncMap           *sync.Map
+	EventCh           chan event.GenericEvent
+}
+
+func (r *StorageAutoscalerScraper) ScrapeMetricsPeriodically(ctx context.Context) {
+	for {
+		select {
+		// context of the main reconcile loop is cancelled
+		case <-ctx.Done():
+			r.Log.Info("context cancelled, stopping scraping metrics")
+			return
+
+		case <-time.After(r.ScrapeInterval):
+			r.enqueueGenericEventsFromMetrics(ctx)
+		}
+	}
+}
+
+func (r *StorageAutoscalerScraper) enqueueGenericEventsFromMetrics(ctx context.Context) {
+	r.Log.Info("scraping metrics")
+	osdUsageQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))"
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdUsageQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return
+	}
+	r.Log.Info("scraped metrics", "metrics", metrics)
+
+	r.Log.Info("update sync map")
+	updateSyncMap(metrics, r.SyncMap)
+	r.Log.Info("sync map updated")
+
+	r.Log.Info("get resource objects from kube api that need scaling")
+	objectList, err := getStorageAutoScalerObjects(ctx, r.Client, r.Log, r.SyncMap)
+	if err != nil {
+		r.Log.Error(err, "failed to get resource objects")
+		return
+	}
```

Comment:
I would suggest separating this into 2 method calls:
1. Get the objects
2. Process the objects from (1.) to figure out which ones need scaling

Unit tests for item (2.) are important for ensuring functionality is correct, and it will be much easier to unit test that function on its own where kube api mocking won't be necessary.